packages/compiler/emitter-framework/asset-emitter.ts

```diff
@@ -0,0 +1,749 @@
+  // 2. the tsp type we're emitting.
+  // 3. the current context.
+  //
+  // Note that in order for this to work, context needs to be interned so
+  // contexts with the same values inside are treated as identical in the
+  // map. See createInterner for more details.
+  const typeToEmitEntity = new CustomKeyMap<[string, Type, ContextState], EmitEntity<T>>(
+    ([method, type, context]) => {
+      return `${method}-${typeId.getKey(type)}-${contextId.getKey(context)}`;
+    }
+  );
+
+  // When we encounter a circular reference, this map will hold a callback
+  // that should be called when the circularly referenced type has completed
+  // its emit.
+  const waitingCircularRefs = new CustomKeyMap<
+    [string, Type, ContextState],
+    {
+      state: EmitterState;
+      cb: (entity: EmitEntity<T>) => EmitEntity<T>;
+    }[]
+  >(([method, type, context]) => {
+    return `${method}-${typeId.getKey(type)}-${contextId.getKey(context)}`;
+  });
+
+  // Similar to `typeToEmitEntity`, this ensures we don't recompute context
+  // for types that we already have context for. Note that context is
+  // dependent on the context of the context call, e.g. if a model is
+  // referenced with reference context set we need to get its declaration
+  // context again. So we use the context's context as a key. Context must
+  // be interned, see createInterner for more details.
+  const knownContexts = new CustomKeyMap<[Type, ContextState], ContextState>(([type, context]) => {
+    return `${typeId.getKey(type)}-${contextId.getKey(context)}`;
+  });
+
+  // The stack of types that the currently emitted type is lexically
+  // contained in. This gets pushed to when we visit a type that is
+  // lexically contained in the current type, and is reset when we jump via
+  // reference to another type in a different lexical context. Note that
+  // this does not correspond to tsp's lexical nesting, e.g. in the case of
+  // an alias to a model expression, the alias is lexically outside the
+  // model, but in the type graph we will consider it to be lexically inside
+  // whatever references the alias.
+  let lexicalTypeStack: Type[] = [];
+
+  // Internally, context is is split between lexicalContext and
+  // referenceContext because when a reference is made, we carry over
+  // referenceContext but leave lexical context behind. When context is
+  // accessed by the user, they are merged by getContext().
+  let context: ContextState = {
+    lexicalContext: {},
+    referenceContext: {},
+  };
+  let programContext: ContextState | null = null;
+  let incomingReferenceContext: Record<string, string> | null = null;
+  const interner = createInterner();
+  const assetEmitter: AssetEmitter<T> = {
+    /**
+     * this
+     * @returns this
+     */
+    getContext() {
+      return {
+        ...context.lexicalContext,
+        ...context.referenceContext,
+      };
+    },
+
+    /**
+     * Return the emitter options
+     */
```

Comment:
Shouldn't these doc comments be moved to the interface so that consumers see it in the IDE?

Reply:
I've deleted these docs... I will make a documentation PR in a subsequent pass (want to iterate on the docs once the compiler docs get checked in).